### PR TITLE
result not unreachable

### PR DESCRIPTION
**Pull request by sbeckeriv**
Dearest Reviewer,

I made a yew based formatter site (https://sbeckeriv.github.io/pest_format/) using a fork of this repo. I dont really want to maintain the the fork. The issue is that I can not figure out how to recover in yew from the unreachable calls. I tried wrapping the call in panic unwind, I tried a layer of indirection. I could not figure it out. I think the code should return a result and let the calling method figure it out. I think this will support more use cases. If anything one can add a format_or_unreachable that unwraps the format and calls unreachable. 

To be clear my use case is to try and format as you type. This does mean lots of cases where you will not be able to reach normally. 

Changes:
Update call unreachable! code to result and cascade the calls.

Thank you for your time on this project.

Dictated but not reviewed,
Becker

### CODE REVIEW

**Comment by sbeckeriv** [comment]
I see you have added a macro for this unreachable.... I will have to update my pr at another time. I think the result is still a better usage.

File: src/formatter.rs [hunk 1]
```diff
@@ -31,10 +31,15 @@ impl Settings {
         let r = read_to_string(path_from)?;
         let s = self.format(&r);
         let mut file = File::create(path_to)?;
-        file.write_all(s.as_bytes())?;
-        return Ok(());
+        match s {
```
Comment (ice1000) [inline src/formatter.rs:34]: Why? Why not do `self.format(&r)?;` at line 32?

File: src/formatter.rs [hunk 2]
```diff
@@ -102,29 +110,34 @@ impl Settings {
...
-                    else if self.choice_first {
-                        let space = std::iter::repeat(' ').take(self.indent - 2).collect::<String>();
-                        code = format!("{{\n  {}}}", indent(&s.join("\n| "), &space));
-                    }
-                    else {
-                        let space = std::iter::repeat(' ').take(self.indent).collect::<String>();
-                        code = format!("{{\n{}}}", indent(&s.join(" |\n"), &space));
+                    match s {
+                        Ok(s) => {
+                            if start == end {
```
Comment (ice1000) [inline src/formatter.rs:115]: Oh please..

**Review by ice1000 — CHANGES_REQUESTED** [review]

File: src/formatter.rs (hunk 2 above)
Comment (sbeckeriv) [reply src/formatter.rs:115]: Hello @ice1000,

I dont understand your comment and would like to learn from this interaction. Could you elaborate on "Oh please.."

Thanks for you time in this review.
Dictated but not reviewed,
Becker

File: src/formatter.rs (hunk 1 above)
Comment (sbeckeriv) [reply src/formatter.rs:34]: Hello @ice1000, 

I did not try it honestly. The signature is for `Result<(), std::io::Error>`. The error from format result is &str.

image

Thanks for your time in the review.

Dictated but not reviewed,
Becker

**Comment by GalAster** [comment]
Should I create a new unified error type?

```rust
#[derive(Debug, Clone)]
pub enum PestError {
    ParseFail(String),
    FormatFail(String),
}
pub type PestResult<T> = Result<T, PestError >;
```

**Comment by sbeckeriv** [comment]
Hello @GalAster,

I do like that idea. It could cover the post comment formatting issue as well https://github.com/pest-parser/pest-fmt/issues/3. 

Thank you again for your time.

Dictated but not reviewed,
Becker

File: src/formatter.rs (hunk 2 above)
Comment (ice1000) [reply src/formatter.rs]: It's also about using `?` instead of pattern matching

File: src/formatter.rs (hunk 1 above)
Comment (ice1000) [reply src/formatter.rs:43]: I guess you can try `.map_err(|e| Error::new(ErrorKind::Other, format!("{}", e)))?`

File: src/formatter.rs (hunk 2 above)
Comment (sbeckeriv) [reply src/formatter.rs]: Hello @ice1000 ,

That is a good suggestion. Had you been clear about your comment the first time I might have committed it.

Thanks
Dictated but not read.
Becker

**Pull request merged**